src/gst-plugins/webrtcendpoint/kmswebrtctransportsrcnice.c

```diff
@@ -29,19 +29,176 @@
+
+#define KMS_WEBRTC_TRANSPORT_SRC_NICE_UNLOCK(src_nice) \
+  (g_rec_mutex_unlock (&(src_nice)->priv->mutex))
+
+#define KMS_WEBRTC_TRANSPORT_SRC_NICE_GET_PRIVATE(obj) ( \
+  G_TYPE_INSTANCE_GET_PRIVATE (                       \
+    (obj),                                            \
+    KMS_TYPE_WEBRTC_TRANSPORT_SRC_NICE,                  \
+    KmsWebrtcTransportSrcNicePrivate                    \
+  )                                                   \
+)
+
+// First byte of a DTLS packet is in the range 19 < B < 64.
+// Doc: https://datatracker.ietf.org/doc/html/rfc5764#section-5.1.2
+#define PACKET_IS_DTLS(b) ((b) > 0x13 && (b) < 0x40)
+
+struct _KmsWebrtcTransportSrcNicePrivate
+{
+  GRecMutex mutex;
+
+  gboolean pending_buffers_delivered;
+
+  GList *pending_buffers;
+};
+
+G_DEFINE_TYPE_WITH_CODE (KmsWebrtcTransportSrcNice, kms_webrtc_transport_src_nice, KMS_TYPE_WEBRTC_TRANSPORT_SRC,
+    G_ADD_PRIVATE (KmsWebrtcTransportSrcNice));
+
+
+
+static gboolean
+gst_buffer_is_dtls (GstBuffer *buffer)
+{
+  guint8 first_byte;
+
+  if (gst_buffer_get_size (buffer) == 0) {
+    GST_DEBUG ("Ignoring buffer with size 0");
+    return FALSE;
+  }
+
+  if (gst_buffer_extract (buffer, 0, &first_byte, 1) != 1) {
+    GST_WARNING ("Could not extract first byte from buffer");
+    return FALSE;
+  }
+
+  return PACKET_IS_DTLS (first_byte);
+}
+
+// Stores DTLS buffers for later use.
+// `buffer` is unref and set to NULL only if stored; otherwise left as-is.
+static gboolean
+store_pending_dtls_buffer (GstBuffer **buffer, guint idx, gpointer user_data)
+{
+  if (gst_buffer_is_dtls (*buffer)) {
+    KmsWebrtcTransportSrcNice *self = KMS_WEBRTC_TRANSPORT_SRC_NICE (user_data);
+
+    GST_DEBUG_OBJECT (self, "Storing DTLS buffer until ICE is CONNECTED");
+    self->priv->pending_buffers = g_list_append (self->priv->pending_buffers, *buffer);
+
+    // Side effect: Unref and remove the buffer from its buffer list, if any.
+    gst_buffer_unref (*buffer);
+    *buffer = NULL;
+  }
+
+  return TRUE;
+}
+
+
 
 static void
 kms_webrtc_transport_src_nice_init (KmsWebrtcTransportSrcNice * self)
 {
   KmsWebrtcTransportSrc *parent = KMS_WEBRTC_TRANSPORT_SRC (self);
 
+  self->priv = KMS_WEBRTC_TRANSPORT_SRC_NICE_GET_PRIVATE (self);
   parent->src = gst_element_factory_make ("nicesrc", NULL);
 
   kms_webrtc_transport_src_connect_elements (parent);
 }
 
+static void
+kms_webrtc_transport_src_nice_finalize (GObject * object)
+{
+  KmsWebrtcTransportSrcNice *self = KMS_WEBRTC_TRANSPORT_SRC_NICE (object);
+
+  if (self->priv->pending_buffers != NULL) {
+    g_list_free_full (self->priv->pending_buffers, (GDestroyNotify)gst_buffer_unref);
+    self->priv->pending_buffers = NULL;
+  }
+}
+
+static void
+kms_webrtc_transport_src_nice_send_pending_buffer (GstBuffer *buffer, KmsWebrtcTransportSrc *self)
+{
+  GstPad *nicesrc_src = gst_element_get_static_pad (self->src, "src");
+
+  if (gst_pad_push (nicesrc_src, buffer) == GST_FLOW_ERROR ) {
+    gst_buffer_unref (buffer);
```

Comment:
No need to unref here, because `gst_pad_push` is transfer-full and we lose our reference:
https://gstreamer.freedesktop.org/documentation/gstreamer/gstpad.html?gi-language=c#gst_pad_push